File: src/wordle.py
import json
from typing import List, Dict


class Wordle:
    def __init__(
        self,
        words_file='data/words.txt',
        freq_file='data/frequency.json',
        info_file='data/info.json'
    ):
        self.SOLUTION: List[str] = []
        self.FREQUENCY: Dict[str, float] = {}
        self.INFO: Dict[str, float] = {}
        self.remaining: List[str] = []
        self.initialize(words_file, freq_file, info_file)

    def initialize(
        self,
        words_file='data/words.txt',
        freq_file='data/frequency.json',
        info_file='data/info.json'
    ):
        # Load solution words
        with open(words_file, 'r') as f:
            self.SOLUTION = [w.strip().lower() for w in f.read().splitlines() if w.strip()]

        # Loading frequency dictionary 
        with open(freq_file, 'r') as f:
            self.FREQUENCY = {k.lower(): v for k, v in json.load(f).items()}

        # Load info dictionary (lowercase keys)
        with open(info_file, 'r') as f:
            self.INFO = {k.lower(): v for k, v in json.load(f).items()}

        # current state
        self.remaining = self.SOLUTION[:]
        self.recompute_available()

    # ---------- helpers ----------
    def recompute_available(self):
        """Recompute availability from remaining words."""
        letters = [set() for _ in range(5)]
        for w in self.remaining:
            for i, ch in enumerate(w):
                letters[i].add(ch)
        self.available = [sorted(list(s)) for s in letters]

    def information(self, s: str) -> float:
        """Sum frequency info of unique letters in a word."""
        info = 0.0
        seen = set()
        for c in s:
            if c not in seen:
                info += self.FREQUENCY.get(c, 0)
                seen.add(c)
        return info

    def isvalid(self, s: str) -> bool:
        """Check if word is in remaining and matches positional availability."""
        if s not in self.remaining:
            return False
        for i, c in enumerate(s):
            if c not in self.available[i]:
                return False
        return True

    # ---------- Wordle feedback ----------
    def feedback(self, secret: str, guess: str) -> str:
        """Return pattern string of length 5 with 'G','Y','B' """
        if secret is None:
            raise TypeError("Did somebody actually called automode without telling the secret word.")

        secret = secret.lower()
        guess = guess.lower()
        res = ['B'] * 5
        used = [False] * 5

        # Greens
        for i in range(5):
            if guess[i] == secret[i]:
                res[i] = 'G'
                used[i] = True

        # Yellows
        for i in range(5):
            if res[i] == 'G':
                continue
            for j in range(5):
                if not used[j] and guess[i] == secret[j]:
                    res[i] = 'Y'
                    used[j] = True
                    break

        return ''.join(res)

    def is_consistent(self, word: str, guess: str, pattern: str) -> bool:
        """Check if 'word' would produce 'pattern' when guessed with 'guess' tbh i just made a one line function idk why."""
        return self.feedback(word, guess) == pattern

    # ---------- state update ----------
    def update(self, guess: str, pattern: str):
        """update remaining based on (guess, pattern) and recompute available."""
        guess = guess.lower()
        pattern = pattern.upper()
        self.remaining = [
            w for w in self.remaining if self.is_consistent(w, guess, pattern)
        ]
        self.recompute_available()

    # ---------- guessing ----------
    def guess(self) -> str:
        """guesses the best word that fits."""
        if not self.remaining:
            return None

        info_scores = {w: self.INFO.get(w, None) for w in self.remaining}
        candidates_with_info = {w: s for w, s in info_scores.items() if s is not None}
        if candidates_with_info:
            return max(candidates_with_info, key=candidates_with_info.get)

        # Fallback 
        best = None
        best_score = (-1, -1.0)
        for w in self.remaining:
            uniq = len(set(w))
            freqsum = sum(self.FREQUENCY.get(c, 0) for c in set(w))
            if (uniq, freqsum) > best_score:
                best_score = (uniq, freqsum)
                best = w
        return best

    # ---------- solver ----------
    def solve(self, secret: str = None, auto: bool = True, verbose: bool = True, max_attempts: int = 6) -> int:
        """
        Play until solved or attempts ended.
        Returns attempts used if solved, else -1.
        """
        # Reset dynamic state
        self.remaining = self.SOLUTION[:]
        self.recompute_available()

        attempts = 0
        while attempts < max_attempts:
            attempts += 1
            g = self.guess()
            if g is None:
                if verbose:
                    print("No valid guess available i guess u lost .")
                return -1

            if verbose:
                print(f"Guess {attempts}: {g}")

            if auto:
                # let it gyess
                if secret is None:
                    raise TypeError("Did somebody actually called automode without telling the secret word.")
                pattern = self.feedback(secret, g)
                if verbose:
                    print(f"Auto feedback: {pattern}")
            else:
                # taking input
                while True:
                    pattern = input("Enter pattern (B/Y/G, length 5): ").strip().upper()
                    if len(pattern) == 5 and all(ch in "BYG" for ch in pattern):
                        break
                    print("Invalid pattern format. Try again.")

            if pattern == "GGGGG":
                if verbose:
                    print(f"Solved in {attempts} attempt(s).")
                return attempts

            prev_len = len(self.remaining)
            self.update(g, pattern)

            if not self.remaining:
                if verbose:
                    print("Feedback is strange so no candidates remain.")
                return -1
            if len(self.remaining) == prev_len and verbose:
                print("No reduction in words so maybe your bot is stupid")

        if verbose:
            print("Attempts ended u lost bro.")
        return -1


if __name__ == "__main__":
    obj = Wordle()
    # Example: auto=False lets you input patterns manually
    print("Result:", obj.solve(auto=False, verbose=True))

File: src/Wordle.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class Wordle {

    public List<String> SOLUTION;
    public Map<String, Double> FREQUENCY;
    public Map<String, Double> INFO;
    public List<String> remaining;
    public List<List<Character>> available;

    public Wordle() throws IOException {
        this("data/words.txt", "data/frequency.json", "data/info.json");
    }

    public Wordle(String words_file, String freq_file, String info_file) throws IOException {
        this.SOLUTION = new ArrayList<>();
        this.FREQUENCY = new HashMap<>();
        this.INFO = new HashMap<>();
        this.remaining = new ArrayList<>();
        initialize(words_file, freq_file, info_file);
    }

    public void initialize(String words_file, String freq_file, String info_file) throws IOException {
        // Load solution words
        for (String w : Files.readAllLines(Paths.get(words_file))) {
            if (w != null && !w.trim().isEmpty()) {
                SOLUTION.add(w.trim().toLowerCase());
            }
        }

        // Loading frequency dictionary 
        parseJson(freq_file, FREQUENCY);

        // Load info dictionary (lowercase keys)
        parseJson(info_file, INFO);

        // current state
        remaining = new ArrayList<>(SOLUTION);
        recompute_available();
    }

    private void parseJson(String file, Map<String, Double> target) throws IOException {
        String json = Files.readString(Paths.get(file)).trim();
        json = json.substring(1, json.length() - 1);
        if (json.isEmpty()) return;
        for (String pair : json.split(",")) {
            String[] kv = pair.split(":");
            String key = kv[0].trim().replace("\"", "").toLowerCase();
            Double val = Double.parseDouble(kv[1].trim());
            target.put(key, val);
        }
    }

    // ---------- helpers ----------
    public void recompute_available() {
        /**Recompute availability from remaining words.*/
        List<Set<Character>> letters = new ArrayList<>();
        for (int i = 0; i < 5; i++) letters.add(new HashSet<>());
        for (String w : remaining) {
            for (int i = 0; i < 5; i++) letters.get(i).add(w.charAt(i));
        }
        available = new ArrayList<>();
        for (Set<Character> s : letters) {
            List<Character> lst = new ArrayList<>(s);
            Collections.sort(lst);
            available.add(lst);
        }
    }

    public double information(String s) {
        /**Sum frequency info of unique letters in a word.*/
        double info = 0.0;
        Set<Character> seen = new HashSet<>();
        for (char c : s.toCharArray()) {
            if (!seen.contains(c)) {
                info += FREQUENCY.getOrDefault(String.valueOf(c), 0.0);
                seen.add(c);
            }
        }
        return info;
    }

    public boolean isvalid(String s) {
        /**Check if word is in remaining and matches positional availability.*/
        if (!remaining.contains(s)) return false;
        for (int i = 0; i < 5; i++)
            if (!available.get(i).contains(s.charAt(i))) return false;
        return true;
    }

    // ---------- Wordle feedback ----------
    public String feedback(String secret, String guess) {
        /**Return pattern string of length 5 with 'G','Y','B' */
        if (secret == null)
            throw new IllegalArgumentException("Did somebody actually called automode without telling the secret word.");

        secret = secret.toLowerCase();
        guess = guess.toLowerCase();
        char[] res = {'B','B','B','B','B'};
        boolean[] used = {false,false,false,false,false};

        // Greens
        for (int i = 0; i < 5; i++) {
            if (guess.charAt(i) == secret.charAt(i)) {
                res[i] = 'G';
                used[i] = true;
            }
        }

        // Yellows
        for (int i = 0; i < 5; i++) {
            if (res[i] == 'G') continue;
            for (int j = 0; j < 5; j++) {
                if (!used[j] && guess.charAt(i) == secret.charAt(j)) {
                    res[i] = 'Y';
                    used[j] = true;
                    break;
                }
            }
        }
        return new String(res);
    }

    public boolean is_consistent(String word, String guess, String pattern) {
        /**Check if 'word' would produce 'pattern' when guessed with 'guess' tbh i just made a one line function idk why.*/
        return feedback(word, guess).equals(pattern);
    }

    // ---------- state update ----------
    public void update(String guess, String pattern) {
        /**update remaining based on (guess, pattern) and recompute available.*/
        guess = guess.toLowerCase();
        pattern = pattern.toUpperCase();
        List<String> next = new ArrayList<>();
        for (String w : remaining)
            if (is_consistent(w, guess, pattern))
                next.add(w);
        remaining = next;
        recompute_available();
    }

    // ---------- guessing ----------
    public String guess() {
        /**guesses the best word that fits.*/
        if (remaining.isEmpty()) return null;

        // 🔒 SAFE entropy usage: valid guesses only, valid outcomes only
        if (remaining.size() <= 500) {
            Map<String, Object> result =
                    Entropy.bestEntropyWord(remaining, remaining);

            String best = (String) result.get("word");
            if (best != null) return best;
        }

        // ---------- fallback: ORIGINAL LOGIC (UNCHANGED) ----------
        String bestInfo = null;
        double bestVal = -1;
        for (String w : remaining) {
            Double v = INFO.get(w);
            if (v != null && v > bestVal) {
                bestVal = v;
                bestInfo = w;
            }
        }
        if (bestInfo != null) return bestInfo;

        // Fallback 
        String best = null;
        int bestUniq = -1;
        double bestFreq = -1;
        for (String w : remaining) {
            Set<Character> uniq = new HashSet<>();
            for (char c : w.toCharArray()) uniq.add(c);
            double freqsum = 0;
            for (char c : uniq)
                freqsum += FREQUENCY.getOrDefault(String.valueOf(c), 0.0);

            if (uniq.size() > bestUniq ||
               (uniq.size() == bestUniq && freqsum > bestFreq)) {
                bestUniq = uniq.size();
                bestFreq = freqsum;
                best = w;
            }
        }
        return best;
    }

    // ---------- solver ----------
    public int solve(String secret, boolean auto, boolean verbose, int max_attempts) {
        /**
        Play until solved or attempts ended.
        Returns attempts used if solved, else -1.
        */
        remaining = new ArrayList<>(SOLUTION);
        recompute_available();

        Scanner sc = auto ? null : new Scanner(System.in);
        int attempts = 0;

        while (attempts < max_attempts) {
            attempts++;
            String g = guess();
            if (g == null) return -1;

            if (verbose) System.out.println("Guess " + attempts + ": " + g);

            String pattern;
            if (auto) {
                if (secret == null)
                    throw new IllegalArgumentException("Did somebody actually called automode without telling the secret word.");
                pattern = feedback(secret, g);
                if (verbose) System.out.println("Auto feedback: " + pattern);
            } else {
                while (true) {
                    System.out.print("Enter pattern (B/Y/G, length 5): ");
                    pattern = sc.nextLine().trim().toUpperCase();
                    if (pattern.length() == 5 && pattern.chars().allMatch(c -> "BYG".indexOf(c) >= 0)) break;
                    System.out.println("Invalid pattern format. Try again.");
                }
            }

            if (pattern.equals("GGGGG")) return attempts;

            int prev = remaining.size();
            update(g, pattern);
            if (remaining.isEmpty()) return -1;
            if (remaining.size() == prev && verbose)
                System.out.println("No reduction in words so maybe your bot is stupid");
        }
        return -1;
    }

    public static void main(String[] args) throws Exception {
        Wordle solver = new Wordle();

        if (args.length >= 1 && args[0].equalsIgnoreCase("hint")) {
            // GUI hint mode: expects remaining words file
            if (args.length < 2) {
                System.err.println("Usage: java Wordle hint <remaining_file>");
                return;
            }
            List<String> remainingWords = Files.readAllLines(Paths.get(args[1]));
            solver.remaining = new ArrayList<>();
            for (String w : remainingWords) if (!w.isBlank()) solver.remaining.add(w.trim().toLowerCase());
            solver.recompute_available();

            String nextGuess = solver.guess();
            System.out.println(nextGuess == null ? "NO_WORDS_LEFT" : nextGuess);
            return;
        }

        // Normal full simulation mode
        long start = System.currentTimeMillis();
        List<String> words = Files.readAllLines(Paths.get("data/words.txt"));
        Map<Integer, Integer> power = new HashMap<>();
        for (int i : new int[]{1,2,3,4,5,6,-1}) power.put(i, 0);

        for (String w : words) {
            if (w.isBlank()) continue;
            int at = solver.solve(w.trim().toLowerCase(), true, false, 6);
            power.put(at, power.get(at) + 1);
        }

        long elapsed = System.currentTimeMillis() - start;

        // prepare output map
        Map<String, Object> output = new HashMap<>();
        output.put("power", power);
        output.put("time", elapsed / 1000.0);

        // print as single line (Python will read it)
        System.out.println(output);
    }

}
